File: php/features.php
<?php
declare(strict_types=1);

// FussyApiDocumentation SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class FussyApiDocumentationFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new FussyApiDocumentationBaseFeature();
            case "ratelimit":
                return new FussyApiDocumentationRatelimitFeature();
            case "retry":
                return new FussyApiDocumentationRetryFeature();
            case "test":
                return new FussyApiDocumentationTestFeature();
            case "timeout":
                return new FussyApiDocumentationTimeoutFeature();
            default:
                return new FussyApiDocumentationBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/fussyapidocumentation_sdk/features.py
# FussyApiDocumentation SDK feature factory

from fussyapidocumentation_sdk.feature.base_feature import FussyApiDocumentationBaseFeature
from fussyapidocumentation_sdk.feature.ratelimit_feature import FussyApiDocumentationRatelimitFeature
from fussyapidocumentation_sdk.feature.retry_feature import FussyApiDocumentationRetryFeature
from fussyapidocumentation_sdk.feature.test_feature import FussyApiDocumentationTestFeature
from fussyapidocumentation_sdk.feature.timeout_feature import FussyApiDocumentationTimeoutFeature


_FEATURES = {
    "base": lambda: FussyApiDocumentationBaseFeature(),
    "ratelimit": lambda: FussyApiDocumentationRatelimitFeature(),
    "retry": lambda: FussyApiDocumentationRetryFeature(),
    "test": lambda: FussyApiDocumentationTestFeature(),
    "timeout": lambda: FussyApiDocumentationTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
